### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by xatapult** [eXist-db/exist#4050]
**Describe the bug**
We have tried to do XML compares using `xmldiff:compare()` and the XPath equivalent `deep-equal()`.

For comparing two equal documents of ~200Kb `xmldiff:compare()` on my machine taks 60msecs, `deep-equal()` over 9 seconds!

**Expected behavior**
The functions  vary slightly in interface, but just comparing two documents should roughly take the same amount of time. And if not 9 seconds is a bit too much I think.

**To Reproduce**
Unzip, import in eXist, execute `test-equals.xql` (I'm doing this from oXygen)

**Context (please always complete the following information):**
 - OS: W10
 - eXist-db version: 5.3
 - Java Version Amazon Corretto 1.8.0_202

**Additional context**
- How is eXist-db installed? Jar
- Any custom changes: various but none seem relevant (mostly external SQL database things)

**Comment by xatapult** [comment]
[TEST.zip](https://github.com/eXist-db/exist/files/7250882/TEST.zip)

**Comment by dizzzz** [comment]
AFAIK for `xmldiff:compare()` two files are serialised and compared with an external tool; I suspect that `deep-equal()` uses a different mechanism and might compare more?

**Comment by line-o** [comment]
I have recently played around with `fn:deep-equal#2` and am wondering what it actually should do.
It does not do well comparing two node-sets from my testing. I ended up serialising both elements and calling deep-equal on that.

**Comment by PieterLamers** [comment]
@dizzzz @adamretter I just heard that there are plans to ditch xmldiff:compare. I'd really hope for this issue to be looked into before this is effectuated.

**Comment by adamretter** [comment]
@PieterLamers I don't think we would intentionally remove it if there is not a good replacement. However, please remind me as we prepare eXist 7, so we don't remove it by accident as part of our tech debt cleanup

**Comment by adamretter** [comment]
@PieterLamers So we definetly are not removing xmldiff:compare - as we just added a new implementation of it here - https://github.com/eXist-db/exist/pull/4554

**Comment by adamretter** [comment]
@xatapult I reran your tests on my laptop:

1. With the new implementation (#4554) of `xmldiff:compare`: 228ms
2. With the new implementation (#4529) of `fn:deep-equals`: 12871ms

It looks like we would still need to optimise `fn:deep-equals`...

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

Returning to this 2021 issue. Built a synthetic reproducer (stored XML, 32k elements with 6 attrs each) and re-measured on develop tip `b917e1ab1d` (eXist 6.x), JDK Zulu 21.38.21:

| Variant | Equal-tree time | Mismatch-tree time |
|---|--:|--:|
| develop, no fix | 2,521 ms | 78 ms |
| With structural-hash short-circuit (Shape α) | 2,676 ms (~6% slower) | 60 ms |

Two findings:

1. **The structural-hash short-circuit doesn't help.** Both `fn:deep-equal` and the proposed hash short-circuit walk each tree exactly once with the same lockstep recursion shape; the hash adds long-allocation + extra accessor calls without skipping any work. Marginally counter-productive on equal trees, marginally helpful on mismatches (but the mismatch path is already fast: 78ms vs 60ms — both dominated by the early-mismatch bail-out).

2. **The dominant cost is per-node persistent-DOM accessor cost** (~79 µs/element). `getFirstChild`, `getNamespaceURI`, `getLocalName`, `getAttributes` each touch the storage layer. Neither a hash short-circuit nor targeted constant-factor fixes (e.g., the "`compareAttributes` is O(attrs²) via `getNamedItemNS`" angle) materially change this — on 6 attrs the squared cost is 36 vs 6 ops, swamped by the per-DOM-call cost.

The 9000ms baseline cited in 2021 may have been closed substantially by intervening refactors (commits 601ba8a74b reducing NPath complexity in `FunDeepEqual` and 82ebd0e21d fixing deep-equal text merging) — my 32k-element synthetic caps at 2.5s on current develop, while the original issue's reported scale is unknown. Without the original `TEST.zip`, I can't confirm whether the 150× ratio still holds.

Suggesting we leave this open pending fresh measurements from the reporter on current develop, or against current `xmldiff:compare`. If the gap is now within ~3-5×, this can probably be closed — the residual is a primitive-level cost (persistent-DOM accessors), not an algorithmic gap fixable at the `fn:deep-equal` layer.

The right *architectural* path to close any remaining gap is Shape β: serialize both subtrees via SAX once per tree and compare the canonical strings. That's how `xmldiff:compare` actually wins — it avoids per-node persistent-DOM round-trips entirely. But it allocates heavily and is a much bigger refactor than Phase 4's surgical-fix discipline supports; would deserve its own design pass and an opt-in `xmldiff:compare`-style fast-path option rather than an unconditional rewrite of `fn:deep-equal`.

Diagnostic harness preserved on a local branch in case fresh timings warrant another look.

**Comment by PieterLamers** [comment]
Hi Joe, thanks for giving this problem a fresh look. In your answer it says 'without the original test.zip': that file is available in the thread. Furthermore, a drawback of the string comparison method is that attribute changes are overlooked by xmldiff:compare.

**Comment by joewiz** [comment]
@PieterLamers Thanks for the pointer, indeed! I'll have Claude take another swing.

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

@PieterLamers Thanks for the pointer — `TEST.zip` retrieved from the thread, `Macbeth.xml` + `Macbeth2.xml` (~3,550 elements each, byte-identical) loaded as stored docs and re-measured on develop tip `b917e1ab1d` (eXist 7.0.0-SNAPSHOT, JDK Zulu 21.38.21).

| Variant | TEST.zip wall time (median of 5) |
|---|--:|
| current develop (no fix) | 5,490 ms |
| with structural-hash short-circuit (Shape α) | 5,482 ms |
| `xmldiff:compare` on the same input (per @adamretter, 2022) | 228 ms |

**Two findings, both reinforcing the prior conclusion:**

1. **Shape α is no help on the actual reproducer either** (within run-to-run noise). The structural-hash short-circuit is doing nothing useful here for the same reason it didn't help on the synthetic 32k tree: hashing visits every node with the same persistent-DOM accessor cost as the existing recursion, so it can't skip work the recursion isn't already skipping.

2. **The original 9000ms baseline is partially closed but still substantial**: TEST.zip went from ~9000ms (eXist 5.3, 2021) → ~5500ms (eXist 7.0.0-SNAPSHOT, 2026). About a 40% improvement from intervening refactors (e.g., `601ba8a74b` `FunDeepEqual` decomposition, `82ebd0e21d` text-merging fix), but the gap to `xmldiff:compare` is still ~24×. Per-element cost on this workload is ~1.55 ms — dominated by per-node persistent-DOM accessor work (`getFirstChild`, `getNamespaceURI`, `getLocalName`, `getAttributes` each touching the storage layer), not by the comparison logic itself.

**On your `xmldiff:compare` attribute caveat**: agreed and noted — that rules out a SAX-serialize-and-compare drop-in replacement (Shape β) for `fn:deep-equal`. `xmldiff:compare`'s "serialize and compare canonical strings" approach gets its speed by avoiding per-node persistent-DOM round-trips entirely, but at the cost of a different correctness profile (no attribute comparison for content equality, different whitespace handling, etc.) — not a behavioural equivalent for `fn:deep-equal`.

**Where this leaves #4050**: the residual gap is real (24× on TEST.zip is not nothing), but closing it requires architectural changes beyond a surgical-fix `fn:deep-equal` patch:
- a SAX-style streaming comparator that respects `fn:deep-equal`'s attribute / namespace / typed-value semantics (substantial design work — would need a custom canonicaliser, not a reuse of existing serialization), or
- in-memory tree caching of stored docs, so `fn:deep-equal` doesn't pay per-node storage-layer cost on each accessor call (broader optimisation, would benefit many other functions besides `fn:deep-equal`).

Both are bigger pieces of work than the current Phase 4 perf-cleanup pass can absorb. Suggesting we leave #4050 open as the tracker for any future deep-equal architecture work, but recognise that the easy wins are exhausted at this layer — the diagnostic harness (`FunDeepEqualPerformanceTest` + `FunDeepEqual4050ReproTest`) and Shape α experiment are preserved on branch `optimize/4050-deep-equal-hash` so a future contributor can pick up either of the architectural paths without re-discovering the hash-short-circuit dead end.
